### ISSUE TITLE
Jetforce is not (ever?) closing sockets

### ISSUE THREAD

**Issue by snoe** [michael-lazar/jetforce#32]
Hi,
I wrote https://github.com/snoe/deedum and I kept running into sites that would hang, one of them was mozz.us, and the other was tilde.black - when the operator of tilde.black said he was also using jetforce I decided to dig into the problem from the server side, since I had pretty much exhausted figuring it out from the client side.

It seems like (perhaps based on some tls configurations or other voodoo) calling `self.transport.loseConnection()` in twisted [does not always lose the connection](https://github.com/twisted/twisted/blob/twisted-17.9.0/src/twisted/protocols/tls.py#L333). However calling `self.transport.transport.loseConnection()` seems to bypass this tls shutdown and properly sends a FIN packet.

Here's two wireshark captures with this difference:
```
diff --git a/jetforce/protocol.py b/jetforce/protocol.py
index dfc53c0..832e305 100644
--- a/jetforce/protocol.py
+++ b/jetforce/protocol.py
@@ -135,7 +135,7 @@ class GeminiProtocol(LineOnlyReceiver):
         finally:
             self.flush_status()
             self.log_request()
-            self.transport.loseConnection()
+            self.transport.transport.loseConnection()

     async def track_deferred(self, deferred: Deferred):
         self._currently_deferred = deferred
```

BEFORE:
<img width="922" alt="transport" src="https://user-images.githubusercontent.com/105012/85686936-a1d5aa80-b68d-11ea-9edd-6ce3269a3e06.png">

AFTER:
<img width="918" alt="transport transport" src="https://user-images.githubusercontent.com/105012/85686947-a4380480-b68d-11ea-9e25-a4524c620b14.png">

**Comment by snoe** [comment]
You can see this behaviour using nmap's `ncat` (don't pipe/redirect stdin or ncat will kill conn from its side): 

```
ncat --ssl -C 127.0.0.1 1965
gemini://192.168.0.196/test.gmi
20 text/gemini
Hello


Ncat: Input/output error.
```

With the patch you'll get the response and the error - without the patch, you get the response only.

**Comment by michael-lazar** [comment]
Hi!

Thanks for your analysis. I was able to reproduce and confirm that your patch does appear to fix the behavior, at least with ncat. I'm no expert in TLS; that's a big reason why I chose to build on top of twisted instead of working with the OpenSSL bindings directly. I want to be cautious with this and make sure that I understand exactly what's going on. I am concerned that dropping the socket without closing TLS cleanly may result in losing untransmitted bytes or something else subtle like that.

This seems related to how the ``close_notify`` alert works (which is painfully not the same between TLS 1.2 and TLS 1.3).

- [RFC 5246 (TLS 1.2)](https://tools.ietf.org/html/rfc5246#section-7.2.1)
- [IETF TLS 1.3](https://tools.ietf.org/id/draft-ietf-tls-tls13-25.html#closure-alerts)

> Either party MAY initiate a close of its write side of the connection by sending a “close_notify” alert. Any data received after a closure alert has been received MUST be ignored. If a transport-level close is received prior to a “close_notify”, the receiver cannot know that all the data that was sent has been received.
> 
> Each party MUST send a “close_notify” alert before closing its write side of the connection, unless it has already sent some other fatal alert. This does not have any effect on its read side of the connection. Note that this is a change from versions of TLS prior to TLS 1.3 in which implementations were required to react to a “close_notify” by discarding pending writes and sending an immediate “close_notify” alert of their own. That previous requirement could cause truncation in the read side. Both parties need not wait to receive a “close_notify” alert before closing their read side of the connection.
> 
> If the application protocol using TLS provides that any data may be carried over the underlying transport after the TLS connection is closed, the TLS implementation MUST receive a “close_notify” alert before indicating end-of-data to the application-layer. No part of this standard should be taken to dictate the manner in which a usage profile for TLS manages its data transport, including when connections are opened or closed.
> 
> Note: It is assumed that closing the write side of a connection reliably delivers pending data before destroying the transport.

I suspect that jetforce calling ``self.transport.loseConnection()`` is actually telling twisted's SSL wrapper to send the ``close_notify`` message. Perhaps the client is never responding with an acknowledgement? Or if it is, I need to be hooking up a callback function somewhere to handle it.

This part in RFC 5246 additionally says:

> If the application protocol will not transfer any additional data, but will only close the underlying transport connection, then the implementation MAY choose to close the transport without waiting for the responding close_notify.

Given this is seems like the best thing to do would be

1. Send the ``close_notify`` alert from the server for the write side of the connection.
2. Tell twisted to destroy the transport while ensuring that the alert has been delivered.

**Comment by snoe** [comment]
That's great you could reproduce, and totally agree that my patch only "works" but there's no understanding behind it. I also do not want to discount ncat and dart both being wrong and maybe twisted is doing the right thing. Unfortunately from flutter/dart - I don't think I can do anything more on my side. Since I don't know the size of the message, I have to rely on reading eof or the socket closing, but I get neither. The tls alert stuff seems to happen somewhere in between that I can't see/access. 



A couple other things I'm trying to figure out on my side, that might be helpful:

I did notice that sending a SHUT_WRITE[1] sometimes caused jetforce to close properly, but also sometimes causes the response channel to be closed[2].

ncat with a `-vvvv` might be useful in the analysis you described but I haven't looked too deeply yet.

I made a command line dart client (hitting gus.guru where I first saw the problem) that might help you:

```dart
import 'dart:io';
import 'dart:convert';

void main(List<String> arguments) async {
var host = 'gus.guru';
  var buf = [];
  var s = await RawSecureSocket.connect(host, 1965, timeout: Duration(seconds: 5));
  var x = s.timeout(Duration(milliseconds: 1500), onTimeout: (x) {
    print("timeout");
    s.close();
    x.close();
  });

  var writeBuf = Utf8Encoder().convert("gemini://" + host + "/" + "\r\n");
  s.setOption(SocketOption.tcpNoDelay, true);
  var wnum = s.write(writeBuf);
  await x.listen((event) {
    switch(event) {
    case RawSocketEvent.read:
      print("read");
            print("r");
            var tbuf = s.read(1024);
            print("tbuf $tbuf");
            buf+=tbuf;
            break;
          case RawSocketEvent.write:
      print("write $wnum ${writeBuf.length}");
            wnum += s.write(writeBuf, wnum);
            if (wnum == writeBuf.length) {

            s.shutdown(SocketDirection.send);
            }
      print("write $wnum");
            break;
          case RawSocketEvent.readClosed:
      print("readClosed");
            break;
          case RawSocketEvent.closed:
      print("close");
            break;
          default:
            throw "Unexpected event $event";

    }

  }).asFuture();

  print(buf.length);
  print(buf);
}
```

I tried a dumb go server that does not exhibit the problem (on port 1966 so I could try it in parallel):

```go
package main

import (
    "crypto/rand"
    "crypto/tls"
    "log"
    "net"
    "crypto/x509"
)

func main() {
    cert, err := tls.LoadX509KeyPair("certs/server.pem", "certs/server.key")
    if err != nil {
        log.Fatalf("server: loadkeys: %s", err)
    }
    config := tls.Config{Certificates: []tls.Certificate{cert}}
    config.Rand = rand.Reader
    service := "0.0.0.0:1966"
    listener, err := tls.Listen("tcp", service, &config)
    if err != nil {
        log.Fatalf("server: listen: %s", err)
    }
    log.Print("server: listening")
    for {
        conn, err := listener.Accept()
        if err != nil {
            log.Printf("server: accept: %s", err)
            break
        }
        defer conn.Close()
        log.Printf("server: accepted from %s", conn.RemoteAddr())
        tlscon, ok := conn.(*tls.Conn)
        if ok {
            log.Print("ok=true")
            state := tlscon.ConnectionState()
            for _, v := range state.PeerCertificates {
                log.Print(x509.MarshalPKIXPublicKey(v.PublicKey))
            }
        }
        go handleClient(conn)
    }
}

func handleClient(conn net.Conn) {
    buf := make([]byte, 512)
        log.Print("server: conn: waiting")
        n, err := conn.Read(buf)
        if err != nil {
            if err != nil {
                log.Printf("server: conn: read: %s", err)
            }
        }
        log.Printf("server: conn: echo %q\n", string(buf[:n]))
        n, err = conn.Write([]byte("20 text/gemini\r\nhello"))
        log.Printf("server: conn: wrote %d bytes", n)

        if err != nil {
            log.Printf("server: write: %s", err)
        }
    log.Println("server: conn: closed")
    conn.Close()
}
```

[1] https://github.com/snoe/deedum/blob/master/lib/net.dart#L169
[2] https://github.com/snoe/deedum/blob/master/lib/net.dart#L67-L69

**Comment by makeworld-the-better-one** [comment]
A lot of this is going over my head, but I just want to chime in and point out that Jetforce is still closing the TLS connection. With clients like [gemget](https://github.com/makeworld-the-better-one/gemget) I can clearly see the difference between a regular page, because the download will complete and the process will stop, versus something that actually never closes like `gemini://chat.mozz.us/stream`.

**Comment by michael-lazar** [comment]
I finally got wireshark working, here are a few more tests (I'm mostly documenting this for my own benefit).

I figured out how to trigger the ``Close Notify`` from python thanks to trio:
https://github.com/python-trio/trio/blob/master/notes-to-self/ssl-close-notify/ssl-close-notify.py

## Scenario 1

1. The client sends the gemini request.
2. The server responds with the content.
3. The server sends ``Close Notify``.
4. The client responds with ``Close Notify``.
5. The server sends ``FIN``.

```python
    with socket.create_connection((host, port)) as sock:
        with context.wrap_socket(sock, server_hostname=sni) as ssock:
            ssock.sendall((url + "\r\n").encode())

            fp = ssock.makefile("rb", buffering=0)
            data = fp.read(1024)
            while data:
                sys.stdout.buffer.write(data)
                sys.stdout.buffer.flush()
                data = fp.read(1024)

            time.sleep(1)

            # Send a close_notify alert
            ssock.setblocking(False)
            ssock.unwrap()

            time.sleep(1)
```

![Screen Shot 2020-07-12 at 12 12 19 AM](https://user-images.githubusercontent.com/5026795/87238737-600a6b00-c3d4-11ea-996a-4b5d2ba832d9.png)


## Scenario 2

1. The client sends the gemini request.
2. The client sends ``Close Notify``.
3. The server responds with the content.
4. The server sends ``Close Notify``.
5. The server sends ``FIN``.

```python
    with socket.create_connection((host, port)) as sock:
        with context.wrap_socket(sock, server_hostname=sni) as ssock:
            ssock.sendall((url + "\r\n").encode())

            # Send a close_notify alert
            ssock.setblocking(False)
            try:
                ssock.unwrap()
            except ssl.SSLWantReadError:
                pass

            time.sleep(5)
```

## Summary

These results confirm my suspicion that twisted's default behavior is to wait for the client to send a ``Close Notify`` before closing the TCP connection. It doesn't matter which order the alerts are sent; the client could send the alert either before or after the server sends the alert. Once the client sends the alert, twisted will immediately send a FIN and close the connection.

**Comment by michael-lazar** [comment]
The OpenSSL shutdown documentation is very helpful, I wish I would have found it earlier!

https://www.openssl.org/docs/man1.1.1/man3/SSL_shutdown.html

I now feel confident enough to update jetforce to always close the connection immediately after sending the close_notify alert.

**Comment by snoe** [comment]
Thank you so much for figuring this out.
